auth/services/selfsigned/signer.go

```diff
@@ -29,66 +29,140 @@ import (
+			// Make sure the session is deleted
+			v.store.Delete(sessionID)
+			return nil, services.ErrSessionNotFound
 		}
-		vp, err = v.vcr.Holder().BuildVP(ctx, []vc.VerifiableCredential{*verifiableCredential}, presentationOptions, &s.issuerDID, true)
+
+		// Create the VerifiablePresentation
+		vp, err = v.createVP(ctx, s)
 		if err != nil {
-			return nil, fmt.Errorf("build VP failed: %w", err)
+			return nil, fmt.Errorf("failed to create VerifiablePresentation: %w", err)
 		}
 	}
 
+	// cleanup all sessions in a final state
+	switch s.Status {
+	case types.SessionVPRequested:
+		fallthrough
+	case types.SessionExpired:
+		fallthrough
+	case types.SessionCancelled:
+		fallthrough
+	case types.SessionErrored:
+		v.store.Delete(sessionID)
+	}
+
 	return signingSessionResult{
 		id:                     sessionID,
-		status:                 s.status,
-		request:                s.contract,
+		status:                 s.Status,
+		request:                s.Contract,
 		verifiablePresentation: vp,
 	}, nil
 }
 
-func (v service) StartSigningSession(rawContractText string, params map[string]interface{}) (contract.SessionPointer, error) {
+// createVP creates a VerifiablePresentation for the given session
+func (v *signer) createVP(ctx context.Context, s types.Session) (*vc.VerifiablePresentation, error) {
+	issuerID, err := did.ParseDID(s.Employer)
+	if err != nil {
+		return nil, fmt.Errorf("invalid issuer DID: %w", err)
+	}
+
+	expirationData := time.Now().Add(24 * time.Hour)
+	credentialOptions := vc.VerifiableCredential{
+		Context:           []ssi.URI{credential.NutsV1ContextURI},
+		Type:              []ssi.URI{ssi.MustParseURI(credentialType)},
+		Issuer:            issuerID.URI(),
+		IssuanceDate:      time.Now(),
+		ExpirationDate:    &expirationData,
+		CredentialSubject: s.CredentialSubject(),
+	}
+	verifiableCredential, err := v.vcr.Issuer().Issue(ctx, credentialOptions, false, false)
+	if err != nil {
+		return nil, fmt.Errorf("issue VC failed: %w", err)
+	}
+	presentationOptions := holder.PresentationOptions{
+		AdditionalContexts: []ssi.URI{credential.NutsV1ContextURI},
+		AdditionalTypes:    []ssi.URI{ssi.MustParseURI(VerifiablePresentationType)},
+		ProofOptions: proof.ProofOptions{
+			Created:      time.Now(),
+			Challenge:    &s.Contract,
+			ProofPurpose: proof.AuthenticationProofPurpose,
+		},
+	}
+	return v.vcr.Holder().BuildVP(ctx, []vc.VerifiableCredential{*verifiableCredential}, presentationOptions, issuerID, true)
+}
+
+func (v *signer) StartSigningSession(userContract contract.Contract, params map[string]interface{}) (contract.SessionPointer, error) {
+	// check the session params first to provide the user with feedback if something is missing
+	if err := checkSessionParams(params); err != nil {
+		return nil, services.InvalidContractRequestError{Message: fmt.Errorf("invalid session params: %w", err)}
+	}
+
+	// TODO: check if the contract name and city matches the employeeDID
```

Comment:
And should check if contract parameters match given parameters?

Reply:
This is checked by the node that issues the access token. But it is still beneficial to do this check preliminary before requesting the access token, since the remote node will probably not give enough information for the local node to find out what went wrong.